source/common/shared_pool/shared_pool.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <thread>
+#include <type_traits>
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/singleton/instance.h"
+
+#include "common/common/assert.h"
+#include "common/common/non_copyable.h"
+#include "common/common/thread_synchronizer.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace SharedPool {
+
+/**
+ * Used to share objects that have the same content.
+ * control the life cycle of shared objects by reference counting
+ *
+ * Note:  ObjectSharedPool needs to be created in the main thread,
+ * all the member methods can only be called in the main thread,
+ * it does not have the ownership of object stored internally, the internal storage is weak_ptr,
+ * when the internal storage object destructor executes the custom deleter to remove its own
+ * weak_ptr from the ObjectSharedPool.
+ *
+ * There is also a need to ensure that the thread where ObjectSharedPool's destructor is also in the
+ * main thread, or that ObjectSharedPool destruct before the program exit
+ */
+template <typename T, typename HashFunc = std::hash<T>,
+          class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
+class ObjectSharedPool : public Singleton::Instance,
+                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
+                         NonCopyable {
+public:
+  ObjectSharedPool(Event::Dispatcher& dispatcher)
+      : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
+
+  void deleteObject(const size_t hash_key) {
+    if (std::this_thread::get_id() == thread_id_) {
+      // There may be new inserts with the same hash value before deleting the old element,
+      // so there is no need to delete it at this time.
+      if (object_pool_.find(hash_key) != object_pool_.end() &&
+          object_pool_[hash_key].use_count() == 0) {
+        object_pool_.erase(hash_key);
+      }
+    } else {
+      // Most of the time, the object's destructor occurs in the main thread, but with some
+      // exceptions, it is destructed in the worker thread. In order to keep the object_pool_ thread
+      // safe, the deleteObject needs to be delivered to the main thread.
+      auto this_shared_ptr = this->shared_from_this();
+      // Used for testing to simulate some race condition scenarios
+      sync_.syncPoint(DeleteObjectOnMainThread);
+      dispatcher_.post([hash_key, this_shared_ptr] { this_shared_ptr->deleteObject(hash_key); });
+    }
+  }
+
+  std::shared_ptr<T> getObject(const T& obj) {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    auto hashed_value = HashFunc{}(obj);
+    auto object_it = object_pool_.find(hashed_value);
```

Comment:
Is hashing sufficient here (in the absence of a crypto strong hash)? Should we have hash followed by an equality comparison?

Reply:
Yes, there is a need to consider hash conflicts. let me open another PR to perfect here.

Reply:
Is this follow-up going to occur? The current code 18 months later is still doing a hash only with no equality comparison.

More concerning is that I see a lot of tsan errors, at least with the toolchain on my workstation. I'm not sure if this is toolchain or hardware related or if something regressed since the original test. I noticed also that docker was used for the original tsan tests and I'm not sure if that makes a difference. I tend to run my tests on RBE without docker these days:
```
bazel test --config=libc++ --config=remote  //test/common/shared_pool:shared_pool_test --config=clang-tsan
...
//test/common/shared_pool:shared_pool_test                               FAILED in 990 out of 1000 in 9.6s
```
I do note that this test passes 1000/1000 times for me when I don't use tsan.

I note also that https://stackoverflow.com/questions/67771015/shared-from-this-and-thread-safety appears conflicted about whether shared_from_this is thread-safe, and it points to https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr#Notes which doesn't appear to address the issue specifically on a quick skim.

Reply:
I'm assuming also that CI is not finding this tsan error generally though I am seeing flakes in coverage tests.

https://gist.github.com/jmarantz/7657d187ad7b011a12b5878be64bc209 is the log of one of my tsan failures.

Reply:
I have a resolution for the tsan test failure -- i think the issue was just that the test thread was racing against the main thread for the destruction of `o2` in 2 tests. https://github.com/envoyproxy/envoy/pull/18262

Reply:
I opened https://github.com/envoyproxy/envoy/issues/24617 to reflect that a follow-up is needed to make this implementation correct in presence of possible hash collisions.